homeassistant/components/sensor/iperf3.py

```diff
@@ -0,0 +1,230 @@
+    vol.Optional(CONF_MINUTE, default=[0]):
+        vol.All(cv.ensure_list, [vol.All(vol.Coerce(int), vol.Range(0, 59))]),
+    vol.Optional(CONF_HOUR):
+        vol.All(cv.ensure_list, [vol.All(vol.Coerce(int), vol.Range(0, 23))]),
+    vol.Optional(CONF_DAY):
+        vol.All(cv.ensure_list, [vol.All(vol.Coerce(int), vol.Range(1, 31))]),
+    vol.Optional(CONF_MANUAL, default=False): cv.boolean,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Iperf3 sensor."""
+    data = Iperf3Data(hass, config)
+
+    dev = []
+    for sensor in config[CONF_MONITORED_CONDITIONS]:
+        dev.append(Iperf3Sensor(data, sensor))
+
+    add_devices(dev)
+
+    def update(call=None):
+        """Update service for manual updates."""
+        data.update(dt_util.now())
+        for sensor in dev:
+            sensor.update()
+
+    for sensor in dev:
+        hass.services.register(DOMAIN, sensor.service_name, update)
+
+
+class Iperf3Sensor(Entity):
+    """A Iperf3 sensor implementation."""
+
+    def __init__(self, iperf3_data, sensor_type):
+        """Initialize the sensor."""
+        self._name = \
+            "{} {}".format(SENSOR_TYPES[sensor_type][0], iperf3_data.server)
+        self._state = None
+        self._sensor_type = sensor_type
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+        self.iperf3_client = iperf3_data
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def service_name(self):
+        """Return the service name of the sensor."""
+        return slugify("{} {}".format(
+            'update_iperf3', self.iperf3_client.server))
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self.iperf3_client.data is not None and \
+           self.iperf3_client.attrs is not None:
+            return self.iperf3_client.attrs
+
+    def update(self):
+        """Get the latest data and update the states."""
+        data = self.iperf3_client.data
+        if data is None:
+            return
+
+        if self._sensor_type == 'download':
+            self._state = data['download']
+        elif self._sensor_type == 'upload':
+            self._state = data['upload']
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        state = yield from async_get_last_state(self.hass, self.entity_id)
+        if not state:
+            return
+        self._state = state.state
+
+    @property
+    def icon(self):
+        """Return icon."""
+        return ICON
+
+
+class Iperf3Data(object):
```

Comment:
Since there is only 1 per entity, why extract it into its own class?

Reply:
Yes, makes sense. I'll simplify it.